File: example-android-project/app/src/main/java/io/github/gofaith/faithdroid/FViews/FSpinner.java
package io.github.gofaith.faithdroid.FViews;

import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Spinner;

import org.json.JSONArray;
import org.json.JSONTokener;

import java.util.ArrayList;
import java.util.List;

import faithdroid.Faithdroid;
import io.github.gofaith.faithdroid.UI.AttrGettable;
import io.github.gofaith.faithdroid.UI.AttrSettable;
import io.github.gofaith.faithdroid.UI.UIController;

public class FSpinner extends FView implements AttrGettable,AttrSettable {
    public Spinner v;
    private List<String> data_list = new ArrayList<>();
    private ArrayAdapter<String> adapter =null;
    private int selected=-1;
    public FSpinner(UIController controller) {
        parentController = controller;
        v = new Spinner(parentController.activity);
        view=v;
        adapter = new ArrayAdapter<>(parentController.activity, android.R.layout.simple_spinner_dropdown_item, data_list);
        v.setAdapter(adapter);
    }
    @Override
    public String getAttr(String attr) {
        String str = getUniversalAttr(attr);
        if (str != null) {
            return str;
        }
        switch (attr){
            // ------------------------------------------
            case "Enabled":
                return String.valueOf(v.isEnabled());
        }
        return "";
    }

    @Override
    public void setAttr(String attr, final String value) {
        if (setUniversalAttr(attr, value)) {
            return;
        }
        switch (attr){
            // -------------------------------------------------------------------
            case "NotifyDataSetChanged":
                adapter.notifyDataSetChanged();
                break;
            case "List":
                try {
                    JSONArray array = (JSONArray) (new JSONTokener(value).nextValue());
                    data_list.clear();
                    for (int i=0;i<array.length();i++) {
                        data_list.add(array.getString(i));
                    }
                    adapter.notifyDataSetChanged();
                } catch (Exception e) {
                    e.printStackTrace();
                }
                break;
            case "ListAdd":
                data_list.add(value);
                adapter.notifyDataSetChanged();
                break;
            case "ListRemove":
                data_list.remove(Integer.parseInt(value));
                adapter.notifyDataSetChanged();
                break;
            case "OnItemClick":
                v.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
                    @Override
                    public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                        Faithdroid.triggerEventHandler(value, String.valueOf(position));
                    }

                    @Override
                    public void onNothingSelected(AdapterView<?> parent) {

                    }
                });
                break;
            case "Enabled":
                v.setEnabled(value.equals("true"));
        }
    }
}

File: android/app/src/main/java/io/github/stevenzack/faithdroidapp/FViews/FSpinner.kt
package io.github.stevenzack.faithdroidapp.FViews

import android.view.View
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.Spinner
import faithdroid.Faithdroid
import io.github.stevenzack.faithdroidapp.UI.AttrGettable
import io.github.stevenzack.faithdroidapp.UI.AttrSettable
import io.github.stevenzack.faithdroidapp.UI.FView
import io.github.stevenzack.faithdroidapp.UI.UIController
import org.json.JSONArray
import org.json.JSONTokener

import java.util.ArrayList

class FSpinner(controller: UIController) : FView(), AttrGettable, AttrSettable {
    var v: Spinner
    private val data_list = ArrayList<String>()
    private var adapter: ArrayAdapter<String>? = null
    private val selected = -1

    init {
        parentController = controller
        v = Spinner(parentController!!.activity)
        view = v
        adapter = ArrayAdapter(parentController!!.activity, android.R.layout.simple_spinner_dropdown_item, data_list)
        v.adapter = adapter
    }

  override  fun getAttr(attr: String): String {
        val str = getUniversalAttr(attr)
        if (str != null) {
            return str
        }
        when (attr) {
            // ------------------------------------------
            "Enabled" -> return v.isEnabled.toString()
            "SelectedIndex" -> return v.selectedItemPosition.toString()
        }
        return ""
    }

   override fun setAttr(attr: String, value: String) {
        if (setUniversalAttr(attr, value)) {
            return
        }
        when (attr) {
            // -------------------------------------------------------------------
            "NotifyDataSetChanged" -> adapter!!.notifyDataSetChanged()
            "List" -> try {
                val array = JSONTokener(value).nextValue() as JSONArray
                data_list.clear()
                for (i in 0 until array.length()) {
                    data_list.add(array.getString(i))
                }
                adapter!!.notifyDataSetChanged()
            } catch (e: Exception) {
                e.printStackTrace()
            }

            "ListAdd" -> {
                data_list.add(value)
                adapter!!.notifyDataSetChanged()
            }
            "ListRemove" -> {
                data_list.removeAt(Integer.parseInt(value))
                adapter!!.notifyDataSetChanged()
            }
            "OnItemClick" -> v.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
                override fun onItemSelected(parent: AdapterView<*>, view: View, position: Int, id: Long) {
                    Faithdroid.triggerEventHandler(value, position.toString())
                }

                override fun onNothingSelected(parent: AdapterView<*>) {

                }
            }
            "Enabled" -> v.isEnabled = value == "true"
        }
    }
}
